python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py

```diff
@@ -0,0 +1,174 @@
+from typing import Callable, Optional, List, Dict, TYPE_CHECKING
+
+import ray
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.data._internal.memory_tracing import trace_allocation
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+)
+from ray.data._internal.execution.util import _merge_ref_bundles
+from ray.data.block import Block, BlockAccessor, BlockExecStats
+from ray.types import ObjectRef
+from ray._raylet import ObjectRefGenerator
+
+if TYPE_CHECKING:
+    from ray.data._internal.execution.operators.map_operator import MapOperator
+
+
+def _map_task(fn: Callable, *blocks: List[Block]):
+    """Remote function for a single operator task.
+
+    Args:
+        fn: The callable that takes Iterator[Block] as input and returns
+            Iterator[Block] as output.
+        blocks: The concrete block values from the task ref bundle.
+
+    Returns:
+        A generator of blocks, followed by the list of BlockMetadata for the blocks
+        as the last generator return.
+    """
+    output_metadata = []
+    stats = BlockExecStats.builder()
+    for b_out in fn(blocks):
+        m_out = BlockAccessor.for_block(b_out).get_metadata([], None)
+        m_out.exec_stats = stats.build()
+        output_metadata.append(m_out)
+        yield b_out
+        stats = BlockExecStats.builder()
+    yield output_metadata
+
+
+class _TaskState:
+    """Tracks the driver-side state for an MapOperator task.
+
+    Attributes:
+        inputs: The input ref bundle.
+        output: The output ref bundle that is set when the task completes.
+    """
+
+    def __init__(self, inputs: RefBundle):
+        self.inputs: RefBundle = inputs
+        self.output: Optional[RefBundle] = None
+
+
+class MapOperatorTasksImpl:
+    def __init__(self, op: "MapOperator"):
+        self._transform_fn = op.get_transform_fn()
+        self._ray_remote_args = op.ray_remote_args()
+        self._tasks: Dict[ObjectRef[ObjectRefGenerator], _TaskState] = {}
+        self._tasks_by_output_order: Dict[int, _TaskState] = {}
+        self._block_bundle = None
```

Comment:
shall we add type hint for `_block_bundle` as well? It would be great to add one-liner comment for what it is as well.